cpp/include/raft/spatial/knn/detail/ann_ivf_flat.cuh

```diff
@@ -0,0 +1,787 @@
+  list_lengths_dev_.resize(nlist_, stream_);
+  centriod_dev_.resize(nlist_ * dim_, stream_);
+
+  // Step 3: Read the list
+  copy(list_prefix_interleaved_dev_.data(), list_prefix_interleaved_host_.data(), nlist_, stream_);
+  copy(list_lengths_dev_.data(), list_lengths_host_.data(), nlist_, stream_);
+  copy(centriod_dev_.data(), centriod_managed_ptr, nlist_ * dim_, stream_);
+
+  copy(list_data_dev_.data(), list_data_host_.data(), ninterleave_ * dim_, stream_);
+  copy(list_index_dev_.data(), list_index_host_.data(), ninterleave_, stream_);
+
+  return cuivflStatus_t::CUIVFL_STATUS_SUCCESS;
+}  // end func cuivflBuildIndex
+
+template <typename T>
+cuivflStatus_t cuivflHandle<T>::queryIVFFlatGridSize(const uint32_t nprobe,
+                                                     const uint32_t batch_size,
+                                                     const uint32_t k)
+{
+  // query the gridDimX size to store probes topK output
+  ivfflat_interleaved_scan<T, typename ivfflat_config<T>::value_t>(nullptr,
+                                                                   nullptr,
+                                                                   nullptr,
+                                                                   nullptr,
+                                                                   nullptr,
+                                                                   nullptr,
+                                                                   metric_type_,
+                                                                   nprobe,
+                                                                   k,
+                                                                   batch_size,
+                                                                   dim_,
+                                                                   nullptr,
+                                                                   nullptr,
+                                                                   stream_,
+                                                                   greater_,
+                                                                   veclen_,
+                                                                   grid_dim_x_);
+  return cuivflStatus_t::CUIVFL_STATUS_SUCCESS;
+}
+
+template <typename T>
+cuivflStatus_t cuivflHandle<T>::cuivflSetSearchParameters(const uint32_t nprobe,
+                                                          const uint32_t max_batch,
+                                                          const uint32_t max_k)
+{
+  nprobe_ = nprobe;
+  if (nprobe_ <= 0) { return CUIVFL_STATUS_INVALID_VALUE; }
+  // Set the greater_
+  if (metric_type_ == raft::distance::DistanceType::L2Expanded ||
+      metric_type_ == raft::distance::DistanceType::L2Unexpanded) {
+    greater_ = false;
+  } else {
+    // Need to set this to true for inner product if need FAISS like behavior for inner product
+    greater_ = false;
+  }
```

Comment:
Answering to other questions, I stumbled upon this today. Would it make sense, from the theory standpoint, to lift `greater_` as an argument of the search function? i.e. something like `ascending` parameter of the topk, to return closest/furthest points. I'm not sure how the "furthest" neighbors mode would perform in terms of precision/recall (since the clustering is based on the nearest neighbours in either case).